### PR TITLE
Updater loop

### PR DESCRIPTION
**Pull request by gabriel**
This fills in some of the updater loop that is available.

The tests here are rudimentary, but we'll be adding more.

@keybase/updater-hackers

### CODE REVIEW

**Comment by chrisnojima** [comment]
@keybase/react-hackers @jzila  bump

File: updater.go [hunk 1]
```diff
@@ -50,8 +56,159 @@ func NewUpdater(source UpdateSource, config Config, log logging.Logger) *Updater
...
-	return nil, nil
+	options := ctx.UpdateOptions()
+	update, err := u.checkForUpdate(ctx, options)
+	if err != nil {
+		return nil, err
+	}
+	if update == nil {
+		// No update available
+		return nil, nil
+	}
```
Comment (jzila) [inline updater.go]: can these two checks be combined into one:
```go
if update == nil {
    return nil, err
}`

File: updater.go [hunk 2]
```diff
@@ -50,8 +56,159 @@ func NewUpdater(source UpdateSource, config Config, log logging.Logger) *Updater
...
+		return nil, findErr
+	}
+	if update == nil {
+		return nil, nil
+	}
+
+	// Save InstallID if we received one
+	if update.InstallID != "" {
+		if err := u.config.SetInstallID(update.InstallID); err != nil {
+			u.log.Warningf("Error saving install ID: %s", err)
```
Comment (jzila) [inline updater.go]: if there's a chance that the error returned can have different types, please use `%#v` to output the error.

File: updater.go [hunk 3]
```diff
@@ -50,8 +56,159 @@ func NewUpdater(source UpdateSource, config Config, log logging.Logger) *Updater
...
+	autoUpdate := !autoSet
+	promptOptions := UpdatePromptOptions{AutoUpdate: autoUpdate}
+	updatePromptResponse, err := updateUI.UpdatePrompt(update, options, promptOptions)
+	if err != nil {
+		return UpdateActionError, err
+	}
+
+	u.log.Debugf("Update prompt response: %#v", updatePromptResponse)
+	if err := u.config.SetUpdateAuto(updatePromptResponse.AutoUpdate); err != nil {
+		u.log.Warningf("Error setting auto preference", err)
```
Comment (jzila) [inline updater.go]: missing error format specifier?

File: updater_osx.go [hunk 4]
```diff
@@ -89,6 +91,35 @@ func (u *Updater) openApplication(applicationPath string) error {
 }
 
 func (u *Updater) platformApplyUpdate(update Update, options UpdateOptions) error {
-	// TODO
+	localPath := update.Asset.LocalPath
+	destinationPath := options.DestinationPath
+	check := func(sourcePath string, destinationPath string) error {
```
Comment (jzila) [inline updater_osx.go]: why is this a closure?

File: updater_test.go [hunk 5]
```diff
@@ -112,3 +116,49 @@ func newDefaultTestUpdateOptions() UpdateOptions {
...
+}
+
+func TestUpdater(t *testing.T) {
+	testServer := testServerForUpdateFile(t, testZipPath)
+	defer testServer.Close()
+
+	upr, err := newTestUpdaterWithServer(t, testServer)
+	assert.NoError(t, err)
+	update, err := upr.Update(testUpdateUI{newDefaultTestUpdateOptions()})
+	if assert.NoError(t, err) {
```
Comment (jzila) [inline updater_test.go]: Hmm, I don't love this branched flow. I prefer to use the `require` package, which just terminates the test in the case that any assertion fails.

**Comment by jzila** [comment]
LGTM after nits.

File: updater.go (hunk 1 above)
Comment (gabriel) [inline updater.go]: I kind of like being really verbose here and following the if err != nil pattern and then saying hey, no error but we don't have an update and that's ok...

If this seems crazy though I can change it to your suggestion.

File: updater.go (hunk 2 above)
Comment (gabriel) [inline updater.go]: Isn't `%#v` outputting go syntax? That seems to be a little awkward for logging given error interface is only `Error() string`?

File: updater.go (hunk 3 above)
Comment (gabriel) [inline updater.go]: Fixed

File: updater_osx.go (hunk 4 above)
Comment (gabriel) [inline updater_osx.go]: Fixed

File: updater.go (hunk 2 above)
Comment (jzila) [inline updater.go]: yes but `%#v` outputs the _type_ of the error.

File: updater.go (hunk 1 above)
Comment (jzila) [inline updater.go]: naw, feel free to leave it.

File: updater_test.go (hunk 5 above)
Comment (gabriel) [inline updater_test.go]: Yes require.. was looking for that. Fixed.

File: updater.go (hunk 2 above)
Comment (gabriel) [inline updater.go]: I have the `Error() string` method returning the internal type in the description.

**Comment by jzila** [comment]
👍

**Pull request merged**